phoenix-core/src/main/java/org/apache/phoenix/query/GuidePostsCache.java

```diff
@@ -85,39 +99,47 @@ public GuidePostsCache(ConnectionQueryServices queryServices, Configuration conf
                 })
                 // Log removals at TRACE for debugging
                 .removalListener(new PhoenixStatsCacheRemovalListener())
-                // Automatically load the cache when entries are missing
-                .build(isStatsEnabled ? new StatsLoader() : new EmptyStatsLoader());
+                // Automatically load the cache when entries need to be refreshed
+                .build(cacheLoader);
     }
 
     /**
-     * {@link CacheLoader} implementation for the Phoenix Table Stats cache.
+     * {@link PhoenixStatsLoader} implementation for the Stats Loader.
      */
-    protected class StatsLoader extends CacheLoader<GuidePostsKey, GuidePostsInfo> {
+    protected class StatsLoaderImpl implements PhoenixStatsLoader {
         @Override
-        public GuidePostsInfo load(GuidePostsKey statsKey) throws Exception {
+        public boolean needsLoad() {
+            // Whenever it's called, we try to load stats from stats table
+            // no matter it has been updated or not.
+            return true;
+        }
+
+        @Override
+        public GuidePostsInfo loadStats(GuidePostsKey statsKey, GuidePostsInfo prevGuidepostInfo) throws Exception {
             @SuppressWarnings("deprecation")
-            Table statsHTable = queryServices.getTable(SchemaUtil.getPhysicalName(
+            TableName tableName = SchemaUtil.getPhysicalName(
                     PhoenixDatabaseMetaData.SYSTEM_STATS_NAME_BYTES,
-                            queryServices.getProps()).getName());
+                    queryServices.getProps());
+            Table statsHTable = queryServices.getTable(tableName.getName());
             try {
                 GuidePostsInfo guidePostsInfo = StatisticsUtil.readStatistics(statsHTable, statsKey,
                         HConstants.LATEST_TIMESTAMP);
                 traceStatsUpdate(statsKey, guidePostsInfo);
                 return guidePostsInfo;
             } catch (TableNotFoundException e) {
                 // On a fresh install, stats might not yet be created, don't warn about this.
-                logger.debug("Unable to locate Phoenix stats table", e);
-                return GuidePostsInfo.NO_GUIDEPOST;
+                logger.debug("Unable to locate Phoenix stats table: " + tableName.toString(), e);
+                return prevGuidepostInfo;
             } catch (IOException e) {
-                logger.warn("Unable to read from stats table", e);
+                logger.warn("Unable to read from stats table: " + tableName.toString(), e);
```

Comment:
Comment from @karanmehta93 (copied from prev PR https://github.com/apache/phoenix/pull/416): nit: remove this comment

Reply:
Why? We need this warning to know some bad thing happened which caused that the stats can't be loaded unexpectedly? This log is printed one time every 15 minutes for each entry only when bad thing is happening, so we won't have too many such logs.

Reply:
I was actually referring to the comment on Line 135 here, which is not valid anymore.

Reply:
Removed the comment.